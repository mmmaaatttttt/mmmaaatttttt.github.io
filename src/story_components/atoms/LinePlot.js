import React from "react";
import PropTypes from "prop-types";
import { line, curveNatural } from "d3-shape";

const LinePlot = ({ graphData, xScale, yScale, stroke, strokeWidth }) => {

  const linePath = line()
    .x(d => xScale(d.x))
    .y(d => yScale(d.y))
    .curve(curveNatural);

  const truncateData = () => graphData.map(d => {
    let newY = d.y;
    const yDomain = yScale.domain();
    if (newY > yDomain[1]) newY = yDomain[1] * 1.1;
    if (newY < yDomain[0]) newY = yDomain[0] * 1.1;
    return { ...d, y: newY };
  });

  return (
    <path
      d={linePath(truncateData())}
      strokeWidth={strokeWidth}
      stroke={stroke}
      fill="none"
    />
  )
};

LinePlot.propTypes = {
  graphData: PropTypes.arrayOf(
    PropTypes.shape({
      x: PropTypes.number.isRequired,
      y: PropTypes.number.isRequired
    })
  ).isRequired,
  xScale: PropTypes.func.isRequired,
  yScale: PropTypes.func.isRequired,
  strokeWidth: PropTypes.string.isRequired,
  stroke: PropTypes.string.isRequired
};

LinePlot.defaultProps = {
  strokeWidth: "5"
};

export default LinePlot;
