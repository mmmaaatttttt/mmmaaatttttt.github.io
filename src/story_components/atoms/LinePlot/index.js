import React from "react";
import PropTypes from "prop-types";
import { line, curveNatural, curveLinear } from "d3-shape";
import { scaleLinear } from "d3-scale";
import COLORS from "utils/styles";

const curves = { curveNatural, curveLinear };

const LinePlot = ({
  curve,
  graphData,
  opacity,
  stroke,
  strokeWidth,
  xScale,
  yScale,
}) => {
  const linePath = line()
    .x(d => xScale(d.x))
    .y(d => yScale(d.y))
    .curve(curves[curve]);

  const truncateData = () =>
    graphData.map(d => {
      let newY = d.y;
      const yDomain = yScale.domain();
      if (newY > yDomain[1]) newY = yDomain[1] * 1.1;
      if (newY < yDomain[0]) newY = yDomain[0] * 1.1;
      return { ...d, y: newY };
    });

  return (
    <path
      d={linePath(truncateData())}
      strokeWidth={strokeWidth}
      stroke={stroke}
      fill="none"
      opacity={opacity}
    />
  );
};

LinePlot.propTypes = {
  curve: PropTypes.oneOf(Object.keys(curves)),
  graphData: PropTypes.arrayOf(
    PropTypes.shape({
      x: PropTypes.number.isRequired,
      y: PropTypes.number.isRequired
    })
  ).isRequired,
  opacity: PropTypes.string.isRequired,
  stroke: PropTypes.string.isRequired,
  strokeWidth: PropTypes.number.isRequired,
  xScale: PropTypes.func.isRequired,
  yScale: PropTypes.func.isRequired
};

LinePlot.defaultProps = {
  curve: "curveNatural",
  graphData: [{ x: 0, y: 0 }, { x: 100, y: 100 }],
  opacity: "1",
  strokeWidth: 5,
  stroke: COLORS.ORANGE,
  xScale: scaleLinear(),
  yScale: scaleLinear()
};

export default LinePlot;
