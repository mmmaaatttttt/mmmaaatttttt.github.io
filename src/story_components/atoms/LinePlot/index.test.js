import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { scaleLinear } from "d3-scale";
import LinePlot from ".";

it("renders successfully", () => {
  shallow(<LinePlot />);
});

it("matches snapshot", () => {
  const wrapper = shallow(
    <LinePlot
      graphData={[
        { x: 10, y: 10 },
        { x: 20, y: 25 },
        { x: 30, y: 40 },
        { x: 50, y: 20 }
      ]}
      xScale={scaleLinear()
        .domain([10, 50])
        .range([0, 200])}
      yScale={scaleLinear()
        .domain([20, 40])
        .range([0, 600])}
      stroke="blue"
      strokeWidth="10"
    />
  );
  expect(toJson(wrapper)).toMatchSnapshot();
});
