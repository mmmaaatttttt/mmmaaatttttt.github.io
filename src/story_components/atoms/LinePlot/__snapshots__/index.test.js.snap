// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<path
  d="M0,60C17.77777777777778,34.66666666666666,35.55555555555556,9.333333333333314,50,150C64.44444444444444,290.6666666666667,75.55555555555556,597.3333333333334,100,600C124.44444444444444,602.6666666666666,162.22222222222223,301.3333333333333,200,0"
  fill="none"
  stroke="blue"
  strokeWidth="10"
/>
`;
